readthedocs/api/v3/mixins.py

```diff
@@ -0,0 +1,110 @@
+from django.contrib.auth.models import User
+from django.shortcuts import get_object_or_404
+from rest_framework.exceptions import NotFound
+
+from readthedocs.builds.models import Version
+from readthedocs.projects.models import Project
+
+
+class NestedParentObjectMixin:
+
+    # Lookup names defined on ``readthedocs/api/v3/urls.py`` when defining the
+    # mapping between URLs and views through the router.
+    PROJECT_LOOKUP_NAMES = [
+        'project__slug',
+        'projects__slug',
+        'superprojects__parent__slug',
+        'main_language_project__slug',
+    ]
+
+    VERSION_LOOKUP_NAMES = [
+        'version__slug',
+    ]
+
+    def _get_parent_object_lookup(self, lookup_names):
+        query_dict = self.get_parents_query_dict()
+        for lookup in lookup_names:
+            value = query_dict.get(lookup)
+            if value:
+                return value
+
+    def _get_parent_project(self):
+        slug = self._get_parent_object_lookup(self.PROJECT_LOOKUP_NAMES)
+        return get_object_or_404(Project, slug=slug)
+
+    def _get_parent_version(self):
+        project_slug = self._get_parent_object_lookup(self.PROJECT_LOOKUP_NAMES)
+        slug = self._get_parent_object_lookup(self.VERSION_LOOKUP_NAMES)
+        return get_object_or_404(
+            Version,
+            slug=slug,
+            project__slug=project_slug,
+        )
+
+
+class APIAuthMixin(NestedParentObjectMixin):
```

Comment:
I think this class could use a better name. It isn't authentication per se, it is a class for getting the correct project queryset given a user and a request. How about ProjectQuerysetMixin?